src/runtime-rs/src/shim/src/task_service.rs

```diff
@@ -0,0 +1,848 @@
+        rsp
+    }
+
+    fn checkpoint(&self, _ctx: &TtrpcContext, _req: shim::CheckpointTaskRequest) -> Result<Empty> {
+        Err(get_rpc_status(
+            Code::NOT_FOUND,
+            "/containerd.task.v2.Task/Checkpoint is not supported".to_string(),
+        ))
+    }
+
+    fn kill(&self, _ctx: &TtrpcContext, req: shim::KillRequest) -> Result<Empty> {
+        debug!("====> kill: {:?}", req);
+        let rsp = self.do_kill(req);
+        debug!("<==== kill: {:?}", rsp);
+        rsp
+    }
+
+    fn exec(&self, _ctx: &TtrpcContext, req: shim::ExecProcessRequest) -> Result<Empty> {
+        debug!("====> exec: {:?}", req);
+        let rsp = self.do_exec(req);
+        debug!("<==== exec: {:?}", rsp);
+        rsp
+    }
+
+    fn resize_pty(&self, _ctx: &TtrpcContext, req: shim::ResizePtyRequest) -> Result<Empty> {
+        debug!("====> resize_pty: {:?}", req);
+        let rsp = self.do_resize_pty(req);
+        debug!("<==== resize_pty: {:?}", rsp);
+        rsp
+    }
+
+    fn close_io(&self, _ctx: &TtrpcContext, req: shim::CloseIORequest) -> Result<Empty> {
+        debug!("====> close_io: {:?}", req);
+        let rsp = self.do_close_io(req);
+        debug!("<==== close_io: {:?}", rsp);
+        rsp
+    }
+
+    fn update(&self, _ctx: &TtrpcContext, req: shim::UpdateTaskRequest) -> Result<Empty> {
+        debug!("====> update: {:?}", req);
+        let rsp = self.do_update(req);
+        debug!("<==== update: {:?}", rsp);
+        rsp
+    }
+
+    fn wait(&self, _ctx: &TtrpcContext, req: shim::WaitRequest) -> Result<shim::WaitResponse> {
+        debug!("====> wait: {:?}", req);
+        let rsp = self.do_wait(req);
+        debug!("<==== wait: {:?}", rsp);
+        rsp
+    }
+
+    fn stats(&self, _ctx: &TtrpcContext, req: shim::StatsRequest) -> Result<shim::StatsResponse> {
+        debug!("====> stats: {:?}", req);
+        let rsp = self.do_stats(req);
+        debug!("<==== stats: {:?}", rsp);
+        rsp
+    }
+
+    fn connect(
+        &self,
+        _ctx: &TtrpcContext,
+        req: shim::ConnectRequest,
+    ) -> Result<shim::ConnectResponse> {
+        debug!("====> connect: {:?}", req);
+        let mut rsp = shim::ConnectResponse::new();
+        rsp.set_shim_pid(self.pid);
+        debug!("<==== connect: {:?}", rsp);
+        Ok(rsp)
+    }
+
+    fn shutdown(&self, _ctx: &TtrpcContext, req: shim::ShutdownRequest) -> Result<Empty> {
+        debug!("====> shutdown: {:?}", req);
+
+        let rsp = Empty::new();
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
+
+        if sandbox.containers.is_empty() {
+            sandbox.try_stop_and_delete();
+            let msg = ServerMessage::new(ServerAction::ShutdownForce);
+            self.shutdown_sender
+                .lock()
+                .expect("poisoned lock")
+                .send(msg)
+                .unwrap();
+        } else {
+            debug!("<==== shutdown: cannot shutdown shim, containers not empty");
```

Comment:
Is there no way to force a shutdown? The code make it look like there should be since the normal case is to specify `ServerAction::ShutdownForce`.

Reply:
This behavior is consistent with the go version kata implementation: do nothing if any container exists. See https://github.com/kata-containers/kata-containers/blob/70274b9d39e8f0ca3f8dcda710510ef8943ab2c0/src/runtime/pkg/containerd-shim-v2/service.go#L936-L943

`ServerAction::ShutdownForce` means to stop listen socket and exit proecss